lib/portlayer/storage/vsphere/volume.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vsphere
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"path"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/lib/portlayer/storage"
+	"github.com/vmware/vic/lib/portlayer/util"
+	"github.com/vmware/vic/pkg/vsphere/disk"
+	"github.com/vmware/vic/pkg/vsphere/session"
+
+	"golang.org/x/net/context"
+)
+
+const volumesDir = "volumes"
+
+// VolumeStore caches Volume references to volumes in the system.
+type VolumeStore struct {
+
+	// maps datastore uri (volume store) to datastore
+	ds map[url.URL]*datastore
+
+	// wraps our vmdks and filesystem primitives.
+	dm *disk.Manager
+
+	sess *session.Session
+}
+
+func NewVolumeStore(ctx context.Context, s *session.Session) (*VolumeStore, error) {
+	dm, err := disk.NewDiskManager(ctx, s)
+	if err != nil {
+		return nil, err
+	}
+
+	v := &VolumeStore{
+		dm:   dm,
+		sess: s,
+		ds:   make(map[url.URL]*datastore),
+	}
+
+	return v, nil
+}
+
+// AddStore adds a volumestore by uri.
+//
+// ds is the Datastore volumes
+// parentDir is the path to parent directory on the datastore.  The path must
+//       exist. The resulting path will be parentDir/VIC/<vch uuid>/volumes.
+// storeName is the name used to refer to the ds + parentDir tupple.
+//
+// returns the URL used to refer to the volume store
+func (v *VolumeStore) AddStore(ctx context.Context, ds *object.Datastore, parentDir, storeName string) (*url.URL, error) {
+	u, err := util.VolumeStoreNameToURL(storeName)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, ok := v.ds[*u]; ok {
+		return nil, fmt.Errorf("volumestore (%s) already added", u.String())
+	}
+
+	// Root our datastore by the directory structured above.
+	p := path.Join(parentDir, storageParentDir, volumesDir)
```

Comment:
Modified this piece of the change since we're getting vic-machine support sooner than expected.